input-events/input-events-typing.html

```diff
@@ -1,82 +1,294 @@
+
+    // Decide whether to use  Key.COMMAND (mac) or Key.CONTROL (everything else)
+    const modifierKey = navigator.platform === "MacIntel" ? '\u2318' : '\uE009';
+
+    // Undo
+    await new test_driver.Actions()
+        .keyDown(modifierKey)
+        .keyDown('z')
+        .keyUp('z')
+        .keyUp(modifierKey)
+        .send();
+    // Redo
+    await new test_driver.Actions()
+        .keyDown(modifierKey)
+        .keyDown('\uE008') // Shift
+        .keyDown('z')
+        .keyUp('z')
+        .keyUp('\uE008')
+        .keyUp(modifierKey)
+        .send();
+
+    // Ignore the initial typing of 'hello'
+    historyInputEventsLog = inputEventsLog.slice(10);
+
+    assert_equals(historyInputEventsLog.length, 4);
+    const inputTypes = ['historyUndo', 'historyRedo'];
+    for (let i = 0; i < historyInputEventsLog.length; i += 2) {
+        // We are increaisng i by 2 as there should always be matching beforeinput and input events.
+        const beforeInputEvent = historyInputEventsLog[i];
+        const inputEvent = historyInputEventsLog[i + 1];
+        assert_equals(beforeInputEvent.type, 'beforeinput');
+        assert_equals(inputEvent.type, 'input');
+        assert_equals(beforeInputEvent.inputType, inputTypes[i / 2]);
+        assert_equals(inputEvent.inputType, inputTypes[i / 2]);
+        assert_equals(beforeInputEvent.data, inputEvent.data);
+    }
+}, 'It triggers beforeinput and input events on typing Undo and Redo key combinations with an existing history');
+
+promise_test(async function() {
+    this.add_cleanup(resetRich);
+    rich.focus();
+    // Decide whether to use  Key.COMMAND (mac) or Key.CONTROL (everything else)
+    const modifierKey = navigator.platform === "MacIntel" ? '\u2318' : '\uE009';
+
+    // Undo
+    await new test_driver.Actions()
+        .keyDown(modifierKey)
+        .keyDown('z')
+        .keyUp('z')
+        .keyUp(modifierKey)
+        .send();
+    // Redo
+    await new test_driver.Actions()
+        .keyDown(modifierKey)
+        .keyDown('\uE008') // Shift
+        .keyDown('z')
+        .keyUp('z')
+        .keyUp('\uE008')
+        .keyUp(modifierKey)
+        .send();
+
+    assert_equals(inputEventsLog.length, 4);
+    const inputTypes = ['historyUndo', 'historyRedo'];
+    for (let i = 0; i < inputEventsLog.length; i += 2) {
+        const beforeInputEvent = inputEventsLog[i];
+        const inputEvent = inputEventsLog[i + 1];
+        assert_equals(beforeInputEvent.type, 'beforeinput');
+        assert_equals(inputEvent.type, 'input');
+        assert_equals(beforeInputEvent.inputType, inputTypes[i / 2]);
+        assert_equals(inputEvent.inputType, inputTypes[i / 2]);
+        assert_equals(beforeInputEvent.data, inputEvent.data);
+    }
+}, 'It triggers beforeinput and input events on typing Undo and Redo key combinations without an existing history');
+
+promise_test(async function() {
+    this.add_cleanup(resetRich);
     const expectedResult = [
```

Comment:
This si the test I didn't write and I inherited how this is done. I am guessing that the reason for why it was done like this is that if one of the results is not what one would expect, it immediately shows in the error message which of the events in which editor the failure can be attributed to.

If we just do:

```
assert_equals(expected[i].type, event.type);
```

And the `type` is not what we expected, then we still need to run the tests a few more times to find out what is "i", so that we find which editor we are talking about and which character was clicked that caused this error.

If you don't think that is a concern, I'll be happy to change those though.

Reply:
If we do further changes, let's do that as a followup.